### ISSUE TITLE
Always generate + output function signatures?

### ISSUE THREAD

**Issue by Gankra** [luser/rust-minidump#336]
TL;DR I think we should unconditionally generate **function** signatures when symbolizing a frame, and the json output should include that value.

-------

# What is a signature

There's kind of 3 different things you can consider a "signature", which I will refer to as:

1. Function signature -- the best possible name we could generate for the "function" at a given address
2. Address signature -- a function signature augmented with more precise position details (module name, source file, source line, instruction offset)
3. Crash signature -- one or more function signatures with additional annotations based on the crash type (crash was due to oom, ipc, js, ...)

Some examples of function signatures:

```
mozilla::TaskController::DoExecuteNextTaskOnlyMainThreadInternal(mozilla::detail::BaseAutoLock<mozilla::Mutex&> const&)
nsAppStartup::Run()
SharedStub
mbae64.dll@0x21f08
0x1142dd57f56e
```

Some examples of address signatures (note the leading `!` and trailing `[...]` used to delimit different parts of the signature):

```
libxul.so!mozilla::TaskController::DoExecuteNextTaskOnlyMainThreadInternal(mozilla::detail::BaseAutoLock<mozilla::Mutex&> const&) [TaskController.cpp:d03f875556391582e06abbf647835af8ca59f94b : 771 + 0x4]
libxul.so!nsAppStartup::Run() [nsAppStartup.cpp:d03f875556391582e06abbf647835af8ca59f94b : 295 + 0x9]
libxul.so!SharedStub + 0x5a
mbae64.dll@0x21f08
0x1142dd57f56e
```

Some examples of crash signatures (note the `|` used to delimit parts of the signature):

```
nsAppStartup::Run()
OOM | small
OOM | large | mozalloc_abort | xul.dll | _PR_NativeRunThread | pr_root
__pthread_kill | pthread_kill | abort | gpusGenerateCrashLog.cold.1
mozilla::detail::MutexImpl::lock | mozilla::layers::NativeLayerRootWayland::AfterFrameClockAfterPaint
```

I believe function signatures are *usually* just the "function name" set by the symbolizer (and exposed in the json), but when the function name isn't available there's some extra logic to come up with a "best-effort alternative" like `mbae64.dll@0x21f08` or `0x1142dd57f56e`, so a user needs of the json or rust-api need to know to do this if they want to properly display a function signature.

In address signatures, note that `[sourcfile: X + Y]` indicates the source line and binary instruction offset, however it's unclear what those values are relative to (module? file? function?). **I actually vaguely recall seeing rust-minidump and socorro disagreeing on this notation, but I never bothered to dig into it since the human-readable output is ostensibly not that important!**

In crash signatures, note that by default the crash signature will be the top frame of the crashing thread, but multiple frames may be included or skipped based on various heuristics.

--------

# Background

While looking at https://bugzilla.mozilla.org/show_bug.cgi?id=1487410 (using machine-readable output for mozcrash.py), I realized I needed to now generate signatures, because the machine-readable output doesn't, but the human does:

https://github.com/luser/rust-minidump/blob/06d2e424c82d68cf263cc0fd9ac2795d043c0ba7/minidump-processor/src/process_state.rs#L334-L366

(Human output generates address signatures but I believe [mozcrash.py parses out the function signature](https://searchfox.org/mozilla-central/rev/6a7c3a1eda4ebb8f9c13779dbbf5eff15bacf8ed/testing/mozbase/mozcrash/mozcrash/mozcrash.py#357-387)!)

This triggered some more serious reflection on our signature generation algorithms, because *ideally* [all our tools would use the same signatures, enabling greater interop](https://bugzilla.mozilla.org/show_bug.cgi?id=867571). This is more difficult than it sounds because socorro uses big frequently-updated [crash signature adjustment config files](https://github.com/willkg/socorro-siggen/tree/main/siggen/siglists), and because it uses information in the [Evil Extra JSON](https://github.com/luser/rust-minidump/blob/master/minidump-processor/src/evil.rs).

Since then two schools of thought have emerged from willkg and I, largely reflecting our domain of expertese:

1. Making a "signatures as a service" web api on symbols.mozilla.org
2. Teaching rust-minidump to generate those signatures (and somehow teach it to pull and evaluate the latest config files).

If I'm being honest, option #1 seems obviously better for Mozilla itself (minimal moving parts, things are trivially in sync, and updates to signature generation that can't be done through config files aren't deployment nightmares).

But this has me thinking about what rust-minidump *can* do for better signature generation, and the pros/cons of doing so.

--------

# In rust-minidump

There are currently 3 places (off the top of my head) in rust-minidump where we output function "names", and therefore where signatures are relevant:

* human output (uses a full **address signature** [algorithm](https://github.com/luser/rust-minidump/blob/06d2e424c82d68cf263cc0fd9ac2795d043c0ba7/minidump-processor/src/process_state.rs#L334-L366))
* machine output (just gives you [the parts you need to synthesize an address signature](https://github.com/luser/rust-minidump/blob/master/minidump-processor/src/process_state.rs#L712-L732))
* stackwalk tracing (does [something quick and dirty to get a bad function signature](https://github.com/luser/rust-minidump/blob/master/minidump-processor/src/stackwalker/mod.rs#L177-L181))

Loose thoughts as I talk this out to myself:

I think always generating and outputting the function signature (which is ostensibly a building block for the other kinds of signature) would be good to do.

On paper, the machine-readable and native-rust outputs are "more important" than the human output, so it's just *weird* that human-readable output has a "feature" that they don't. However consumers of machine-output probably want a function signature, and not an address signature? We could emit both but that's kinda weird. 

Since the function signature is *usually* just the "function name", perhaps we should fill in missing function names *with* the function signature? That way we wouldn't be emitting the same value twice in the common case. Not sure if that would confuse anything that wants to know if there was a "real" function name.

This would also make the machine-readable output more generally readable and useful to a non-power user (one who has no notion of how to generate any kind of signatures). Similarly the native-rust interface would be nicer to a non-power user.

Function (or address?) signatures would also be nice for the tracing logs, but I went for something blunt and simple to try to keep the cost of logging down (especially because i'm not 100% sure how expression evaluation inside a conditionally-printing macro like `trace!` works). If we always *had* to generate them, then it would be "free" to use them in logging.

I am not *certain* of this fact, but my intuition is that socorro's **crash** signature generation only ever needs to concatenate function signatures, and that function signature generation is rarely-if-ever updated. If so, then it is viable to at least move that logic into rust-minidump (and regardless it needs function signature generation to render the backtraces in its UI). To my knowledge socorro does not use address signatures at all.

Also if we're being honest generating a few more format strings is probably completely trivial to our execution time but death by a thousand cuts with useless features can be a thing.

**Comment by willkg** [comment]
re: "function signatures are rarely--if ever--updated": Socorro's signature generation normalizes function names before using them for signature generation. I adjust that periodically to handle issues with functions that don't match across platforms (and probably other things, too). I know there are a few outstanding bugs with it now that I haven't gotten to, yet.

re: "crash signatures": I'm currently throwing together a web api. I might get it done by the end of next week--we'll see.

re: "address signatures": I don't think Socorro or Socorro users use address signatures. I don't think we generate or store them, but it's possible Socorro users can generate them from the data we do have. If they're interesting, I'm game for looking into adding them.

**Comment by Gankra** [comment]
Could you point me to the code that does function signature normalization? That sounds very interesting.

**Comment by willkg** [comment]
It's here:

https://github.com/mozilla-services/socorro/blob/40750d53e594f96481120b3c203fc327ea4dcf42/socorro/signature/rules.py#L95-L230

I think it's super gross, but I haven't had a chance to rethink it. Rethinking how that's implemented is one of the things I want to do before I rework signature generation for Java crashes.
